### DISCUSSION TITLE
Pinecil boot logo flashing

### DISCUSSION THREAD

**Discussion by crc-32** [Ralim/IronOS#742] [General]
Didn't see any info on how to do this, correct me if it's somewhere I've not found or any of the steps are unnecessary, but I managed to flash a boot logo with the following:

 - `python3 img2ts100.py input.png logo.hex`
 - `riscv-nuclei-elf-objcopy -I ihex -O binary logo.hex logo.bin`
 - `dfu-util -d 28e9:0189 -a 0 -D logo.bin -s 0x0800f800`

I would add this to the wiki but I don't think just anyone has access, and the repo docs just point there.

**Comment by terencode** [comment]
How did you do to flash the firmware file? 
I used 
`sudo dfu-util -d 28e9:0189 -a 0 -D Pinecil.hex -s 0x08000000:mass-erase:force`

but while the image seems to have been flashed correctly, it doesn't boot.

**Comment by Ralim** [comment]
Ah, yes I will need to document this somewhere 

The first command is correct, you need to flash it without the mass-erase after flashing the normal firmware.

If you include the mass-erase you will erase the firmware!

You can reflash using the latest bin file from the master branch artefacts to get the firmware back.

**Reply by terencode** [reply]
so your firmware is overwriting the original firmware starting from the beginning up to a certain address?  How does this work?

> You can reflash using the latest bin file from the master branch artefacts to get the firmware back.

Thanks but I had a backup :)

Tomorrow or later I will update the pinecil wiki to account for that.

**Reply by terencode** [reply]
If I remove `:mass-erase:force` I get 
```
Downloading element to address = 0x08000000, size = 136112
dfu-util: Last page at 0x080213af is not writeable
```

**Reply by Ralim** [reply]
So the device has 128k of flash.
The firmare is setup to only use 126k of that at most (Currently its around 60k)
It has "reserved" the last two pages of flash for the settings and the logo.
So what you are doing is a partial flash update to flash the logo to the right page, after flashing the firmware.
`mass-erase` causes _all_ flash to be erased, but what we want is to just write to the last page.

The address 0xF800 is where it was located in the firmware currently, which is actually artificially limiting firmware size and I will bump it out in a later release. [Here is where that address turns up.](https://github.com/Ralim/ts100/blob/22c31cb27d57ee9ea7b258d6d131d77e242a38f1/workspace/TS100/Core/BSP/Pine64/logo.cpp#L11)

**Reply by terencode** [reply]
Ah but I'm talking about flashing the firmware I generated using `./build.sh -l FR -m Pinecil`.

**Reply by Ralim** [reply]
Ah  that's different.

In which case you should be able to flash the binary at the 0x8000000 with mass erase:force

**Reply by crc-32** [reply]
From my testing I don't think dfu-util takes hex files though (they are just considered binaries and soft brick the iron) so be sure to flash the *binary* not the hex file, which is likely why it didn't boot the first time you tried

**Reply by terencode** [reply]
> From my testing I don't think dfu-util takes hex files though (they are just considered binaries and soft brick the iron) so be sure to flash the _binary_ not the hex file, which is likely why it didn't boot the first time you tried

I'll try right now, I assumed it was .hex  based on the file Ralim provided at https://github.com/Ralim/ts100/issues/730#issuecomment-744083576

EDIT: Yup it's working :)

**Reply by crc-32** [reply]
I'm not sure, but it didn't work for me until I converted it

**Reply by terencode** [reply]
What's the interest behind providing the .hex and .elf in addition to the .bin ?

**Reply by crc-32** [reply]
The compiler produces the .elf, it's then converted to a .bin and then I assume it's also converted to a .hex as the ts100 is able to use that file format, so regardless of what target you're using it produces that file too.

**Comment by terencode** [comment]
I can confirm the boot image flashing works: I modified the TS100 picture to look more like the pinecil and then uploaded it, here is the result:


https://user-images.githubusercontent.com/12665563/103175025-40261100-4867-11eb-8819-41625722becb.mov

**Comment by Ralim** [comment]
Noting I did throw this in [here](https://github.com/Ralim/IronOS/blob/master/Documentation/Logo.md)
And put flashing docs [here](https://github.com/Ralim/IronOS/blob/master/Documentation/Flashing.md)
